archunit/src/main/java/com/tngtech/archunit/core/importer/ClassGraphCreator.java

```diff
@@ -154,14 +156,61 @@ private void completeTypeParametersAndMembers() {
     }
 
     private void completeAnnotations() {
+        ensureAnnotationsHierarchy();
         for (JavaClass javaClass : classes.getAllWithOuterClassesSortedBeforeInnerClasses()) {
             DomainObjectCreationContext.completeAnnotations(javaClass, this);
             for (JavaMember member : concat(javaClass.getFields(), javaClass.getMethods(), javaClass.getConstructors())) {
-                memberDependenciesByTarget.registerAnnotations(member.getAnnotations());
+                resolveAndCompleteMemberAnnotations(member);
             }
         }
     }
 
+    private void ensureAnnotationsHierarchy() {
+        for (String annotation : importRecord.getAllAnnotationNames()) {
+            resolveAnnotations(annotation);
+        }
+    }
+
+    private void resolveAnnotations(String annotation) {
+        JavaClass javaClassOfAnnotation = classes.getOrResolve(annotation);
+        for (JavaAnnotationBuilder metaAnnotation : importRecord.getAnnotationsFor(javaClassOfAnnotation.getName())) {
+            String metaAnnotationName = metaAnnotation.getTypeDescriptor().getFullyQualifiedClassName();
+            if (isAlreadyImported(metaAnnotationName)) {
+                break;
+            }
+            resolveAnnotations(metaAnnotationName);
+        }
+    }
+
+    private boolean isAlreadyImported(String annotation) {
+        for (JavaClass javaClass : classes.getAllWithOuterClassesSortedBeforeInnerClasses()) {
+            if (javaClass.getFullName().equals(annotation)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void resolveAndCompleteMemberAnnotations(JavaMember member) {
+        memberDependenciesByTarget.registerAnnotations(member.getAnnotations());
+        Set<String> resolvedAnnotations = new HashSet<>();
+        for (JavaAnnotation<? extends JavaMember> annotation : member.getAnnotations()) {
+            resolveMemberAnnotations(annotation, resolvedAnnotations);
+        }
+    }
+
+    private void resolveMemberAnnotations(JavaAnnotation<?> annotation, Set<String> resolvedAnnotations) {
+        DomainObjectCreationContext.completeAnnotations(annotation.getRawType(), this);
+        resolvedAnnotations.add(annotation.getRawType().getName());
+        for (JavaAnnotation<JavaClass> metaAnnotation : annotation.getRawType().getAnnotations()) {
+            if (resolvedAnnotations.contains(metaAnnotation.getRawType().getName())) {
```

Comment:
I think this has the same problem as I addressed in the one review commit, breaking the whole loop as soon as any already resolved annotation name is encountered.
But I'm also wondering why for class meta-annotations you simply use `isAlreadyImported() -> ImportedClasses` while here you need your custom set of visited annotations? And if there is a reason to use this set here, instead of `ImportedClasses`, why is the scope limited to a single member? Do I really need to reset the set for a second member annotated with the same annotation type? I would think not? :thinking: